resources: Add doc for riscv-fs-busybox-opensbi-nodisk resource

teloecho opened the pull request:
Inspired by riscv-fs-nodisk however, replacing legacy Berkeley bootloader (bbl) with OpenSBI FW_PAYLOAD bootloader, additionally just using default Linux and Busybox configurations instead of not up-to-date UCanLinux configs.

I was missing an up-to-date guide for fully self-build simple linux environment for Gem5 RISC-V full system simulations.
Therefore i compiled this guide from other guide here and online together.
The use case could be playing around with custom RISC-V instructions and custom toolchain.

Btw: i tried to use pre-commit, but i got the error `No .pre-commit-config.yaml file was found` on commit.

teloecho committed:
resources: Add riscv-fs/riscv-busybox-opensbi

Inspired by riscv-fs-nodisk however, replacing legacy Berkeley
bootloader (bbl) with OpenSBI FW_PAYLOAD bootloader, additionally
just using default Linux and Busybox configurations instead of not
up-to-date UCanLinux configs.

teloecho commented:
@powerjg or @BobbyRBruce i would like to add you as reviewers, but i am not sure, if i am authorized.

powerjg commented on src/riscv-fs/riscv-busybox-opensbi/README.md:
````diff
@@ -0,0 +1,249 @@
+found,
+```sh
+export PATH=/path/to/rvtoolchain/bin/:$PATH
+```
````
Do we have to compile the toolchain? If there's a good reason not to use the precompiled toolchain from Ubuntu, then this step is OK, but if it's possible to use the precompiled toolchain, then we should do that.

teloecho replied:
It might be helpful if you modify the toolchain itself, however in the default case for this resource the shipped toolchain is faily enough, so i changed it to use Debian's `g++-12-riscv64-linux-gnu` package. See [lines 43-74](https://github.com/gem5/gem5-resources/blob/2cb2179cacf52fa62aaa19c64315849d6a96787c/src/riscv-fs/riscv-busybox-opensbi/README.md?plain=1#L43-L73).

powerjg commented on src/riscv-fs/riscv-busybox-opensbi/README.md:
```diff
@@ -0,0 +1,249 @@
+git clone --branch 1_36_stable https://git.busybox.net/busybox.git # choose a stable branch
+
+# alternatively downloading tar
+wget https://git.busybox.net/busybox/snapshot/busybox-1_36_1.tar.bz2
```
I assume we *do* have to compile busybox (unlike gcc above). But can you confirm this?

teloecho replied:
Yes, Busybox is not distributing for the required RISCV platform (see their [download page](https://busybox.net/downloads/binaries/)). Also i configure it to use no shared libraries in order to simplify the guide.

powerjg commented on src/riscv-fs/riscv-busybox-opensbi/README.md:
```diff
@@ -0,0 +1,249 @@
+cd linux
+# ?? CFLAGS_KERNEL=" -march=rv64gc"
+make ARCH=riscv CROSS_COMPILE=riscv64-unknown-linux-gnu- defconfig
+make ARCH=riscv CROSS_COMPILE=riscv64-unknown-linux-gnu- menuconfig
```
`riscv64-unknown-linux-gnu-`? Above you used `riscv64-linux-gnu-`.

We should use just one cross compiler, and, unless there's a really good reason, it should be a precompiled toolchain, not one that we compile ourselves.

teloecho replied:
In the initial commit the [lines that referred to `riscv64-linux-gnu-`](https://github.com/gem5/gem5-resources/blob/ef7547f07caebbd58e7743cf7a13c559232dfd56/src/riscv-fs/riscv-busybox-opensbi/README.md?plain=1#L95-L100) just mentioned an alternative usage if the toolchain was *not* prefixed by the default `riscv64-unknown-linux-gnu-`.  Sorry for the confusion.

As i use the Debian shipped compilers now, i have to specify the toolchain prefix for `scons` anyways. See [line 109](https://github.com/gem5/gem5-resources/blob/2cb2179cacf52fa62aaa19c64315849d6a96787c/src/riscv-fs/riscv-busybox-opensbi/README.md?plain=1#L109).

powerjg commented on src/riscv-fs/riscv-busybox-opensbi/README.md:
````diff
@@ -0,0 +1,249 @@
+You can run the created bootloader with Linux and Busybox payload using the example riscv `fs_linux.py` config,
+```sh
+cd gem5
+./build/RISCV/gem5.opt configs/example/riscv/fs_linux.py --kernel="../gem5-resources/src/riscv-fs/riscv-all-in-one/opensbi/build/platform/generic/firmware/fw_payload.elf"
````
I would prefer all examples to use the stdlib. If you could describe how to turn this into a `Workload` it would improve the description.

teloecho replied:
This is the hard part for me. As far as i see, only specifying a bootloader without kernel and disk image is not intended by the boards, e.g. [RiscvBoard](https://www.gem5.org/documentation/general_docs/stdlib_api/gem5.components.boards.riscv_board.html) or [RISCVMatchedBoard](https://www.gem5.org/documentation/general_docs/stdlib_api/gem5.prebuilt.riscvmatched.riscvmatched_board.html).

Therefore i just show how to slightly modify the [`riscv-fs.py`](https://github.com/gem5/gem5/blob/stable/configs/example/gem5_library/riscv-fs.py) by obtaining the new `kernel` or `bootloader` resource and leaving the `disk_image` (and `kernel`) parameter of the `set_kernel_disk_workload` call unmodified (as dummy resource). See [lines 266-282](https://github.com/gem5/gem5-resources/blob/2cb2179cacf52fa62aaa19c64315849d6a96787c/src/riscv-fs/riscv-busybox-opensbi/README.md?plain=1#L266-L282)

In the end it might look like
```python
# Set the Full System workload.
board.set_kernel_disk_workload(
    kernel=obtain_resource(
        "riscv-bootloader-vmlinux-5.10", resource_version="1.0.0"
    ), # dummy resource
    disk_image=obtain_resource("riscv-disk-img", resource_version="1.0.0"),  # dummy resource
    bootloader=obtain_resource(
        "riscv-fs-busybox-opensbi-nodisk", resource_version="1.0.0"
    ),
)
```

powerjg reviewed (changes requested):
First, thanks for the contribution! I think this information is great to get out to the broader community.

However, I'm not sure the appropriate place is gem5-resources. This repo is for the sources to build resources that are available on resources.gem5.org. Are you proposing that we (the gem5 community) distribute this kernel/bootloader/filesystem? If so, who are the users of this?

Maybe a better place for this document would be the gem5 blog: https://www.gem5.org/blog/

I'm open to a conversation, though :)

One last thing, if you could run a markdown linter, it would improve the formatting. E.g., https://github.com/markdownlint/markdownlint or https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint if you're using vscode.

teloecho committed:
resources: convert to precompiled toolchain version

following suggestion of review on PR #61 to use the more convinient
precompiled toolchain. self-compiled is useful for modified toolchain,
but not required for reproducibility of the standard resource though

teloecho committed:
resources: formatting and clarifications

improvements for PR #61
- fix markdown formatting (blank lines, links, etc.)
- clarify comment alternative initramfs.cpio archive generation

teloecho committed:
resources: turn into bootloader resource, improve metadata

- change permalink and add a tag to highlight nodisk charakter
- reorder tags to align with other resource READMEs
- describe how to use the resource within the gem5 stdlib framework
  as a bootloader or kernel passed to set_kernel_disk_workload()
- add licensing notes for OpenSBI, Linux and Busybox

teloecho commented:
Thank you very much for the thorough review and the hints for improvement!

I should have incorporated your suggestions somehow (see comments on your comments :).

> However, I'm not sure the appropriate place is gem5-resources. This repo is for the sources to build resources that are available on resources.gem5.org. Are you proposing that we (the gem5 community) distribute this kernel/bootloader/filesystem? If so, who are the users of this?
> 
> Maybe a better place for this document would be the gem5 blog: https://www.gem5.org/blog/
> 
> I'm open to a conversation, though :)

Regarding the appropriate place.
Well, i would propose that you distribute this bootloader resource indeed.
Although i was/am not sure either.
The advantage of this resource is its straight forward reproducibility, usability for many cases and small size.
It might be useful if you like to craft a workload utilizing systemcalls or run several user programs and if you like to observe the whole boot process.
Utilizing custom instructions relevant for the boot process usually requires to rebuild a lot from toolchain via bootloader and kernel up to the user binaries. These instructions might guide you very well through all the steps then.

So why to distribute via gem5-resources? -- It might be a good starting point for people to check whether this type of "nodisk resource" already fits their needs.
Would this convince you? If not, i would also be fine with a blog post.

teloecho committed:
resources: align path names

teloecho committed:
resources: update to version ref to new gem5 release

teloecho committed:
resources: fix init script to actually exit

- move m5 exit to beginning of script and extract shell setup into
  extra script for better readability
- swap setsid,hotplug lines (script doesn't proceed after setsid line)
- fix version in busybox docs link
